--- a/configure.ac
+++ b/configure.ac
@@ -171,8 +171,7 @@
 	(lirc_dev lirc_parallel) \
 	(lirc_dev lirc_sasem) \
 	(lirc_dev lirc_serial) \
-	(lirc_dev lirc_sir) \
-	(lirc_dev lirc_wpc8769l)"
+	(lirc_dev lirc_sir)"
 fi
 
 AH_TEMPLATE([DAEMONIZE],
@@ -445,7 +444,7 @@
                           srm7500libusb, tekram,
                           tekram_bt829, tira, tira_raw, ttusbir,
                           tuxbox, tvbox, udp, uirt2, uirt2_raw,
-                          usb_uirt_raw, usbx, wpc8769l],
+                          usb_uirt_raw, usbx],
 driver=${withval},
 driver="unset"
 )
@@ -608,8 +607,6 @@
 		;;
 	lirc_dev-lirc_sir)
 		;;
-	lirc_dev-lirc_wpc8769l)
-		;;
 	lirc_flyvideo)
 		;;
 	livedrive_midi)
@@ -1506,8 +1503,7 @@
 	lirc_sasem \
 	lirc_serial \
 	lirc_sir \
-	lirc_ttusbir \
-	lirc_wpc8769l"
+	lirc_ttusbir"
 fi
 if kernel_module=$(expr "$lirc_driver" : 'lirc_dev \(.*\)'); then
   :
