--- a/tools/rbp/setup-sdk.sh
+++ b/tools/rbp/setup-sdk.sh
@@ -2,7 +2,7 @@
 
 SCRIPT_PATH=$(cd `dirname $0` && pwd)
 
-USE_BUILDROOT=1
+USE_BUILDROOT=0
 
 if [ "$USE_BUILDROOT" = "1" ]; then
   BUILDROOT=/opt/xbmc-bcm/buildroot
@@ -15,9 +15,9 @@
 else
   TARBALLS=/opt/xbmc-tarballs
   SDKSTAGE=/opt/bcm-rootfs
-  XBMCPREFIX=/opt/xbmc-bcm/xbmc-bin
+  XBMCPREFIX=/usr/local
   TARGETFS=/opt/bcm-rootfs
-  TOOLCHAIN=/usr/local/bcm-gcc
+  TOOLCHAIN=/toolchain/arm-bcm2708hardfp-linux-gnueabi
   BUILDROOT=/opt/bcm-rootfs
 fi
 
--- a/tools/rbp/depends/depends.mk
+++ b/tools/rbp/depends/depends.mk
@@ -4,7 +4,7 @@
 	export BUILD=i686-linux
 	export PREFIX=$(XBMCPREFIX)
 	export SYSROOT=$(BUILDROOT)/output/host/usr/arm-unknown-linux-gnueabi/sysroot
-	export CFLAGS+=-isystem$(SYSROOT)/usr/include -isystem$(SYSROOT)/opt/vc/include -isystem$(SDKSTAGE)/opt/vc/include/interface/vcos/pthreads -isystem$(PREFIX)/include -isystem$(PREFIX)/usr/include/mysql --sysroot=$(SYSROOT) -isystem$(SDKSTAGE)/opt/vc/include/interface/vmcs_host/linux/
+	export CFLAGS+=-isystem$(SYSROOT)/usr/include -isystem$(SYSROOT)/opt/vc/include -isystem$(SDKSTAGE)/opt/vc/include/interface/vcos/pthreads -isystem$(PREFIX)/include -isystem$(PREFIX)/usr/include/mysql --sysroot=$(SYSROOT)
 	export CXXFLAGS=$(CFLAGS) --sysroot=$(SYSROOT)
 	export CPPFLAGS=$(CFLAGS) --sysroot=$(SYSROOT)
 	export LDFLAGS=-L$(SYSROOT)/opt/vc/lib -L$(XBMCPREFIX)/lib
@@ -22,19 +22,19 @@
 	export NATIVE_ROOT=/usr
 	export PYTHON_LDFLAGS=-L$(SDKSTAGE)/usr/lib -lpython$(PYTHON_VERSION) -lpthread -ldl -lutil -lm
 else
-	export CFLAGS=-pipe -O3 -mcpu=arm1176jzf-s -mtune=arm1176jzf-s -mfloat-abi=softfp -mfpu=vfp -mabi=aapcs-linux -Wno-psabi -Wa,-mno-warn-deprecated -Wno-deprecated-declarations
-	export HOST=arm-bcm2708-linux-gnueabi
+	export CFLAGS=-pipe -O3 -mcpu=arm1176jzf-s -mtune=arm1176jzf-s -mfloat-abi=hard -mfpu=vfp -mabi=aapcs-linux -Wno-psabi -Wa,-mno-warn-deprecated -Wno-deprecated-declarations
+	export HOST=arm-bcm2708hardfp-linux-gnueabi
 	export BUILD=i686-linux
 	export PREFIX=${XBMCPREFIX}
 	export TARGETFS
-	export SYSROOT=/usr/local/bcm-gcc/arm-bcm2708-linux-gnueabi/sys-root
-	export RLINK_PATH=-Wl,-rpath-link,${SYSROOT}/lib -Wl,-rpath-link,${TARGETFS}/lib -Wl,-rpath-link,${TARGETFS}/usr/lib -Wl,-rpath-link,${TARGETFS}/opt/vc/
-	export CFLAGS+=-isystem${XBMCPREFIX}/include -isystem${SDKSTAGE}/usr/include -isystem${SDKSTAGE}/opt/vc/include -isystem$(SDKSTAGE)/opt/vc/include/interface/vcos/pthreads -isystem${SDKSTAGE}/opt/vc -isystem$(SDKSTAGE)/opt/vc/include/interface/vmcs_host/linux/
-	export CFLAGS+=-L${XBMCPREFIX}/lib -L${SYSROOT}/lib -L${TARGETFS}/lib -L${TARGETFS}/usr/lib -L${TARGETFS}/opt/vc/lib ${RLINK_PATH}
+	export SYSROOT=/toolchain/arm-bcm2708hardfp-linux-gnueabi/arm-bcm2708hardfp-linux-gnueabi/sys-root
+	export RLINK_PATH=-Wl,-rpath-link,${SYSROOT}/lib -Wl,-rpath-link,${TARGETFS}/lib/arm-linux-gnueabihf:${TARGETFS}/usr/lib/arm-linux-gnueabihf:${TARGETFS}/lib:${TARGETFS}/usr/lib -Wl,-rpath-link,${TARGETFS}/opt/vc/lib
+	export CFLAGS+=-isystem${XBMCPREFIX}/include -isystem${SDKSTAGE}/usr/include/arm-linux-gnueabihf -isystem${SDKSTAGE}/usr/include -isystem${SDKSTAGE}/usr/local/include -isystem${SDKSTAGE}/opt/vc/include -isystem$(SDKSTAGE)/opt/vc/include/interface/vcos/pthreads -isystem${SDKSTAGE}/opt/vc/include/interface/vmcs_host -isystem${SDKSTAGE}/opt/vc/include/interface/vmcs_host/linux
+	export CFLAGS+=-L${XBMCPREFIX}/lib -L${SYSROOT}/lib -L${SYSROOT}/usr/lib -L${TARGETFS}/lib -L${TARGETFS}/usr/lib -L${TARGETFS}/usr/local/lib -L${TARGETFS}/opt/vc/lib ${RLINK_PATH} -L${TARGETFS}/usr/lib/arm-linux-gnueabihf -L${TARGETFS}/lib/arm-linux-gnueabihf
 	export CXXFLAGS=${CFLAGS}
 	export CPPFLAGS=${CFLAGS}
-	export LDFLAGS=${RLINK_PATH} -L${TARGETFS}/lib -L${TARGETFS}/usr/lib -L${XBMCPREFIX}/lib
-	export LD=${TOOLCHAIN}/bin/${HOST}-ld
+	export LDFLAGS=${RLINK_PATH} -L${SYSROOT}/usr/lib -L${TARGETFS}/lib -L${TARGETFS}/usr/lib -L${TARGETFS}/usr/local/lib -L${TARGETFS}/opt/vc/lib -L${XBMCPREFIX}/lib -L${TARGETFS}/usr/lib/arm-linux-gnueabihf -Wl,--dynamic-linker=/lib/ld-linux-armhf.so.3
+ 	export LD=${TOOLCHAIN}/bin/${HOST}-ld
 	export AR=${TOOLCHAIN}/bin/${HOST}-ar
 	export CC=${TOOLCHAIN}/bin/${HOST}-gcc
 	export CXX=${TOOLCHAIN}/bin/${HOST}-g++
@@ -43,9 +43,9 @@
 	export STRIP=${TOOLCHAIN}/bin/${HOST}-strip
 	export OBJDUMP=${TOOLCHAIN}/bin/${HOST}-objdump
 	#export ACLOCAL=aclocal -I ${SDKSTAGE}/usr/share/aclocal -I ${TARGETFS}/usr/share/aclocal-1.11
-	export PKG_CONFIG_LIBDIR=${PREFIX}/lib/pkgconfig:${SDKSTAGE}/lib/pkgconfig:${SDKSTAGE}/usr/lib/pkgconfig
-	export PKG_CONFIG_PATH=$(PREFIX)/bin/pkg-config
-	export PYTHON_VERSION=2.6
+	export PKG_CONFIG_LIBDIR=${PREFIX}/lib/pkgconfig:${SDKSTAGE}/lib/pkgconfig:${SDKSTAGE}/usr/lib/pkgconfig:${SDKSTAGE}/usr/local/lib/pkgconfig:${SDKSTAGE}/usr/lib/arm-linux-gnueabihf/pkgconfig
+ 	export PKG_CONFIG_PATH=$(PREFIX)/bin/pkg-config
+	export PYTHON_VERSION=2.7
 	export PATH:=${PREFIX}/bin:$(PATH):${TOOLCHAIN}/bin
 	export NATIVE_ROOT=/usr
 	export PYTHON_LDFLAGS=-L$(SDKSTAGE)/usr/lib -lpython$(PYTHON_VERSION)
--- a/tools/rbp/depends/xbmc/Makefile
+++ b/tools/rbp/depends/xbmc/Makefile
@@ -9,7 +9,7 @@
 
 CONFIGURE=./configure --prefix=$(PREFIX) --build=$(BUILD) --host=$(HOST) \
 	--enable-gles --disable-sdl --disable-x11 --disable-xrandr --disable-openmax \
-	--disable-optical-drive --disable-dvdcss --disable-joystick --disable-debug \
+	--enable-optical-drive --enable-dvdcss --enable-joystick --disable-debug --enable-libbluray \
 	--disable-crystalhd --disable-vtbdecoder --disable-vaapi --disable-vdpau \
 	--disable-pulse --disable-projectm --with-platform=raspberry-pi --enable-optimizations \
 	--enable-libcec --enable-player=omxplayer
