--- MediaLibrary.js.old	2013-05-12 14:25:05.100708433 +0000
+++ MediaLibrary.js	2013-05-12 14:28:24.572805011 +0000
@@ -259,18 +259,18 @@
           });
         case 'forward':
           return xbmc.rpc.request({
-            'method': 'Player.SetSpeed',
+            'method': 'Player.Seek',
             'params': {
               'playerid': player,
-              'speed': 'increment'
+              'value': 'smallforward'
             }
           });
         case 'rewind':
           return xbmc.rpc.request({
-            'method': 'Player.SetSpeed',
+            'method': 'Player.Seek',
             'params': {
               'playerid': player,
-              'speed': 'decrement'
+              'value': 'smallbackward'
             }
           });
       }
